libbpf-tools/cpudist.c

```diff
@@ -0,0 +1,314 @@
+		const char *format, va_list args)
+{
+	if (level == LIBBPF_DEBUG && !env.verbose)
+		return 0;
+	return vfprintf(stderr, format, args);
+}
+
+static int bump_memlock_rlimit(void)
+{
+	struct rlimit rlim_new = {
+		.rlim_cur	= RLIM_INFINITY,
+		.rlim_max	= RLIM_INFINITY,
+	};
+
+	return setrlimit(RLIMIT_MEMLOCK, &rlim_new);
+}
+
+static int get_pid_max(void)
+{
+	int pid_max;
+	FILE *f;
+
+	f = fopen("/proc/sys/kernel/pid_max", "r");
+	if (!f)
+		return -1;
+	if (fscanf(f, "%d\n", &pid_max) != 1)
+		pid_max = -1;
+	fclose(f);
+	return pid_max;
+}
+
+static void sig_handler(int sig)
+{
+	exiting = true;
+}
+
+static void pid_to_comm(pid_t pid, char *comm)
+{
+	char fname[64];
+	FILE *f;
+
+	snprintf(fname, sizeof(fname), "/proc/%d/comm", pid);
+	f = fopen(fname, "r");
+	if (!f || fscanf(f, "%s\n", comm) < 0)
+		*comm = '\0';
+	if (f)
+		fclose(f);
+}
+
+static int print_log2_hists(int fd)
+{
+	char *units = env.milliseconds ? "msecs" : "usecs";
+	__u32 lookup_key = -2, next_key;
+	struct hist hist;
+	char comm[16];
+	int err;
+
+	while (!bpf_map_get_next_key(fd, &lookup_key, &next_key)) {
+		err = bpf_map_lookup_elem(fd, &next_key, &hist);
+		if (err < 0) {
+			fprintf(stderr, "failed to lookup hist: %d\n", err);
+			return -1;
+		}
+		pid_to_comm(next_key, comm);
+		if (env.per_process)
+			printf("\npid = %d %s\n", next_key, comm);
+		if (env.per_thread)
+			printf("\ntid = %d %s\n", next_key, comm);
+		print_log2_hist(hist.slots, MAX_SLOTS, units);
+		lookup_key = next_key;
+	}
+
+	lookup_key = -2;
+	while (!bpf_map_get_next_key(fd, &lookup_key, &next_key)) {
+		err = bpf_map_delete_elem(fd, &next_key);
+		if (err < 0) {
+			fprintf(stderr, "failed to cleanup hist : %d\n", err);
+			return -1;
+		}
+		lookup_key = next_key;
+	}
+	return 0;
+}
+
+int main(int argc, char **argv)
+{
+	static const struct argp argp = {
+		.options = opts,
+		.parser = parse_arg,
+		.doc = argp_program_doc,
+	};
+	struct cpudist_bpf *obj;
+	int pid_max, fd, err;
+	struct tm *tm;
+	char ts[32];
+	time_t t;
+
+	err = argp_parse(&argp, argc, argv, 0, NULL, NULL);
+	if (err)
+		return err;
+
+	libbpf_set_print(libbpf_print_fn);
+
+	err = bump_memlock_rlimit();
+	if (err) {
+		fprintf(stderr, "failed to increase rlimit: %d\n", err);
+		return 1;
+	}
+
+	obj = cpudist_bpf__open();
+	if (!obj) {
+		fprintf(stderr, "failed to open and/or load BPF ojbect\n");
+		return 1;
+	}
+
+	/* initialize global data (filtering options) */
+	obj->rodata->targ_per_process = env.per_process;
+	obj->rodata->targ_per_thread = env.per_thread;
+	obj->rodata->targ_ms = env.milliseconds;
+	obj->rodata->targ_offcpu = env.offcpu;
+	obj->rodata->targ_tgid = env.pid;
+
+	pid_max = get_pid_max();
+	if (pid_max < 0) {
+		fprintf(stderr, "failed to get pid_max\n");
+		return 1;
+	}
+
+	bpf_map__resize(obj->maps.start, pid_max);
+	if (!env.per_process && !env.per_thread)
+		bpf_map__resize(obj->maps.hists, 1);
+	else
+		bpf_map__resize(obj->maps.hists, pid_max);
```

Comment:
Hi @anakryiko For hists, if we changed it to ARRAY, will the traversal overhead in user space be relatively large when we only have hundred threads? 

I simply compared overhead with `syscount`. Based on the map type, the bpf system call is only called a few hundred times , and it will be called 8w times after changing to ARRAY during the same time period.

Reply:
True. You can add BPF_F_MMAPABLE flag to ARRAY and mmap() it. Then accessing it won't require any syscalls and will be extremely fast. But I'm not going to ask you to do this. Let's stick to HASHMAP for now.

Reply:
Cool, I'll learn more about `BPF_F_MMAPABLE` + mmap later then use it in next tools. Thank you.